src/cdk/scrolling/for-of.ts

```diff
@@ -0,0 +1,275 @@
+})
+export class CdkForOf<T> implements CollectionViewer, DoCheck, OnDestroy {
+  /** Emits when the rendered view of the data changes. */
+  viewChange = new Subject<Range>();
+
+  /** Emits when the data source changes. */
+  private _dataSourceSubject = new Subject<DataSource<T>>();
+
+  /** The DataSource to display. */
+  @Input()
+  get cdkForOf(): NgIterable<T> | DataSource<T> { return this._cdkForOf; }
+  set cdkForOf(value: NgIterable<T> | DataSource<T>) {
+    this._cdkForOf = value;
+    let ds = value instanceof DataSource ? value :
+        new ArrayDataSource<T>(Array.prototype.slice.call(value));
+    this._dataSourceSubject.next(ds);
+  }
+  _cdkForOf: NgIterable<T> | DataSource<T>;
+
+  /** The trackBy function to use for tracking elements. */
+  @Input()
+  get cdkForTrackBy(): TrackByFunction<T> {
+    return this._cdkForOfTrackBy;
+  }
+  set cdkForTrackBy(fn: TrackByFunction<T>) {
+    this._needsUpdate = true;
+    this._cdkForOfTrackBy =
+        (index, item) => fn(index + (this._renderedRange ? this._renderedRange.start : 0), item);
+  }
+  private _cdkForOfTrackBy: TrackByFunction<T>;
+
+  /** The template used to stamp out new elements. */
+  @Input()
+  set cdkForTemplate(value: TemplateRef<CdkForOfContext<T>>) {
+    if (value) {
+      this._needsUpdate = true;
+      this._template = value;
+    }
+  }
+
+  /** Emits whenever the data in the current DataSource changes. */
+  dataStream: Observable<T[]> = this._dataSourceSubject
+      .pipe(
+          startWith(null!),
+          pairwise(),
+          switchMap(([prev, cur]) => this._changeDataSource(prev, cur)),
+          shareReplay(1));
+
+  private _differ: IterableDiffer<T> | null = null;
+
+  private _data: T[];
+
+  private _renderedItems: T[];
+
+  private _renderedRange: Range;
+
+  private _templateCache: EmbeddedViewRef<CdkForOfContext<T>>[] = [];
+
+  private _needsUpdate = false;
+
+  constructor(
+      private _viewContainerRef: ViewContainerRef,
+      private _template: TemplateRef<CdkForOfContext<T>>,
+      private _differs: IterableDiffers,
+      @Host() private _viewport: CdkVirtualScrollViewport) {
+    this.dataStream.subscribe(data => this._data = data);
+    this._viewport.renderedRangeStream.subscribe(range => this._onRenderedRangeChange(range));
+    this._viewport.connect(this);
+  }
+
+  /**
+   * Get the client rect for the given index.
+   * @param index The index of the data element whose client rect we want to measure.
+   * @return The combined client rect for all DOM elements rendered as part of the given index.
+   *     Or null if no DOM elements are rendered for the given index.
+   * @throws If the given index is not in the rendered range.
+   */
+  measureClientRect(index: number): ClientRect | null {
+    if (index < this._renderedRange.start || index >= this._renderedRange.end) {
+      throw Error('Error: attempted to measure an element that isn\'t rendered.');
+    }
+    index -= this._renderedRange.start;
```

Comment:
Introduce a new variable instead of modifying a function argument? `renderedIndex`?